File: rhode.py
#!usr/bin/python
#coding=utf-8

import sys
reload(sys)
sys.setdefaultencoding('utf8')
import os
import cv2
import time
from glob import glob

pyPath = os.path.dirname(os.path.realpath(__file__))

if not os.path.isdir(os.path.join(pyPath, 'target')):
	os.makedirs(os.path.join(pyPath, 'target'))
if not os.path.isdir(os.path.join(pyPath, 'stop')):
	os.makedirs(os.path.join(pyPath, 'stop'))
if not os.path.isdir(os.path.join(pyPath, 'template')):
	os.makedirs(os.path.join(pyPath, 'template'))

def findPos(template, target):
	target_rgb = cv2.imread(target)
	t_h, t_w = target_rgb.shape[0:2]
	target_gray = cv2.cvtColor(target_rgb, cv2.COLOR_BGR2GRAY)
	template_rgb = cv2.imread(template, 0)
	res = cv2.matchTemplate(target_gray, template_rgb, cv2.TM_CCOEFF_NORMED)
	value = cv2.minMaxLoc(res)
	rate = value[1]
	print rate
	if rate > 0.9:
		return (value[-1][0] + (t_w / 2), value[-1][1] + (t_h / 2))
	return None

def execute(cmd):
	print cmd
	os.system(cmd)

def click(device_id, pos):
	cmd = 'adb -s %s shell input tap %s %s'%(device_id, pos[0], pos[1])
	execute(cmd)

def screenshot(device_id, tag):
	cmd = 'adb -s %s shell screencap -p /sdcard/rhode_%s.jpg'%(device_id, tag)
	execute(cmd)
	cmd = 'adb -s %s pull /sdcard/rhode_%s.jpg %s/template/rhode_%s.jpg'%(device_id, tag, pyPath, tag)
	execute(cmd)
	return '%s/template/rhode_%s.jpg'%(pyPath, tag)

def getDeviceId():
	cmd = 'adb devices'
	out = os.popen(cmd).read()
	deviceList = out.split('\n')
	deviceList = filter(lambda x: x != '' and 'List of devices' not in x, [i.split('\tdevice')[0].strip() for i in deviceList])
	return deviceList[0]

def lock(device_id):
	cmd = 'adb -s %s shell input keyevent 26'%device_id
	execute(cmd)

def main():
	device_id = getDeviceId()
	print 'android: %s'%device_id
	target_list = glob('target/*.jpg')
	target_list.sort()
	stop_list = glob('stop/*.jpg')
	stop_list.sort()
	close = False
	while not close:
		for tar in target_list:
			template = screenshot(device_id, 'template')
			print tar
			pos = findPos(template, tar)
			while pos is None:
				template = screenshot(device_id, 'template')
				print tar
				pos = findPos(template, tar)
				time.sleep(3)
			click(device_id, pos)
			time.sleep(3)
			stop = screenshot(device_id, 'stop')
			for st in  stop_list:
				ret = findPos(stop, st)
				if ret:
					close = True
					break
			if close:
				break
	lock(device_id)

if __name__ == '__main__':
	main()

File: rhode.go
package main

import (
	"fmt"
	"os"
	"sort"
	"time"
	"os/exec"
	"io/ioutil"
	"gocv.io/x/gocv"
	"strings"
)

func PathExists(path string, makedir bool) bool {
	isExist := false
	_, err := os.Stat(path)
	if err == nil {
		isExist = true
	}
	if os.IsNotExist(err) {
		isExist = false
	}
	if !isExist && makedir {
		os.Mkdir(path, os.ModePerm)
	}
	return isExist
}

func GetAllFile(pathname string, s []string, ext string) ([]string, error) {
	rd, err := ioutil.ReadDir(pathname)
	if err != nil {
		fmt.Println("read dir fail:", err)
		return s, err
	}
	for _, fi := range rd {
		if fi.IsDir() {
			fullDir := pathname + "/" + fi.Name()
			s, err = GetAllFile(fullDir, s, ext)
			if err != nil {
				fmt.Println("read dir fail:", err)
				return s, err
			}
		} else {
			fullName := pathname + "/" + fi.Name()
			if strings.HasSuffix(fullName, ext) {
				s = append(s, fullName)
			}
		}
	}
	return s, nil
}

func Typeof(v interface{}) string {
	return fmt.Sprintf("%T", v)
}

func execute(cmd string, args []string) {
	fmt.Println(fmt.Sprintf("%+v %+v", cmd, strings.Join(args, " ")))
	exec.Command(cmd, args...)
}

func findPos(template string, target string) []int {
	target_gray := gocv.IMRead(target, gocv.IMReadGrayScale)
	template_gray := gocv.IMRead(template, gocv.IMReadGrayScale)
	t_h, t_w := target_gray.Size()[0], target_gray.Size()[1]
	ret := gocv.NewMat()
	defer ret.Close()
	gocv.MatchTemplate(target_gray, template_gray, &ret, gocv.TmCcoeffNormed, gocv.NewMat())
	_, rate, _, pos := gocv.MinMaxLoc(ret)
	fmt.Println(rate)
	if rate > 0.9 {
		return []int{pos.X + (t_w / 2), pos.Y + (t_h / 2)}
	}
	return nil
}

func getDeviceId() string {
	cmd := exec.Command("adb", "devices")
	buf, _ := cmd.Output()
	out := string(buf)
	deviceList := strings.Split(out, "\n")
	deviceId := ""
	for _, v := range deviceList{
		if !strings.Contains(v, "List of devices") && strings.TrimSpace(v) != "" {
			deviceId = strings.Split(v, "\tdevice")[0]
		}
	}
	return strings.TrimSpace(deviceId)
}

func screenshot(device_id string, tag string) string {
	args := strings.Split(fmt.Sprintf("-s %+v shell screencap -p /sdcard/rhode_%+v.jpg", device_id, tag), " ")
	execute("adb", args)
	args = strings.Split(fmt.Sprintf("-s %+v pull /sdcard/rhode_%+v.jpg template/rhode_%+v.jpg", device_id, tag, tag), " ")
	execute("adb", args)
	return fmt.Sprintf("template/rhode_%+v.jpg", tag)
}

func click(device_id string, pos []int) {
	args := strings.Split(fmt.Sprintf("-s %+v shell input tap %+v %+v", pos[0], pos[1]), " ")
	execute("adb", args)
}

func lock(device_id string) {
	args := strings.Split(fmt.Sprintf("-s %+v shell input keyevent 26", device_id), " ")
	execute("adb", args)
}

func main() {
	PathExists("target", true)
	PathExists("stop", true)
	PathExists("template", true)
	device_id := getDeviceId()
	fmt.Println(fmt.Sprintf("android: %+v", device_id))
	target_list, _ := GetAllFile("target", make([]string, 0), ".jpg")
	sort.Strings(target_list)
	stop_list, _ := GetAllFile("stop", make([]string, 0), ".jpg")
	sort.Strings(stop_list)
	exit := false
	for {
		if exit {
			break
		}
		for _, tar := range target_list {
			template := screenshot(device_id, "template")
			fmt.Println(tar)
			pos := findPos(template, tar)
			if len(pos) == 2 {
				click(device_id, pos)
				time.Sleep(time.Duration(5) * time.Second)
			}
		}
		stop := screenshot(device_id, "stop")
		for _, st := range stop_list {
			ret := findPos(stop, st)
			if len(ret) == 2 {
				exit = true
				break
			}
		}
	}
	lock(device_id)
}